src/stream/src/executor/merge.rs

```diff
@@ -464,6 +477,92 @@ impl SelectReceivers {
     }
 }
 
+/// A wrapper that buffers the `StreamChunk`s from upstream until no more ready items are available.
+/// Besides, any message other than `StreamChunk` will trigger the buffered `StreamChunk`s
+/// to be emitted immediately, as well as the message itself.
+struct BufferChunks<S: Stream> {
+    inner: S,
+    chunk_builder: StreamChunkBuilder,
+
+    /// The items to be emitted. Whenever there's something here, we should return a `Poll::Ready` immediately.
+    pending_items: VecDeque<S::Item>,
+}
+
+impl<S: Stream> BufferChunks<S> {
+    pub(super) fn new(inner: S, chunk_size: usize, schema: Schema) -> Self {
+        assert!(chunk_size > 0);
+        let chunk_builder = StreamChunkBuilder::new(chunk_size, schema.data_types());
+        Self {
+            inner,
+            chunk_builder,
+            pending_items: VecDeque::new(),
+        }
+    }
+}
+
+impl<S: Stream> std::ops::Deref for BufferChunks<S> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<S: Stream> std::ops::DerefMut for BufferChunks<S> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<S: Stream> Stream for BufferChunks<S>
+where
+    S: Stream<Item = std::result::Result<Message, StreamExecutorError>> + Unpin,
+{
+    type Item = std::result::Result<Message, StreamExecutorError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(item) = self.pending_items.pop_front() {
+                return Poll::Ready(Some(item));
+            }
+
+            match self.inner.poll_next_unpin(cx) {
+                Poll::Pending => {
+                    return if self.chunk_builder.size() > 0 {
+                        let chunk_out = self.chunk_builder.take().unwrap();
+                        Poll::Ready(Some(Ok(Message::Chunk(chunk_out))))
+                    } else {
+                        Poll::Pending
+                    }
+                }
+
+                Poll::Ready(Some(result)) => {
+                    if let Ok(Message::Chunk(chunk)) = result {
+                        for row in chunk.records() {
+                            if let Some(chunk_out) = self.chunk_builder.append_record(row) {
+                                self.pending_items.push_back(Ok(Message::Chunk(chunk_out)));
+                            }
+                        }
+                    } else {
+                        return if self.chunk_builder.size() > 0 {
+                            let chunk_out = self.chunk_builder.take().unwrap();
```

Comment:
I think we can just use `self.chunk_builder.take()` and do not need to check the `chunk_builder.size() > 0`.

Reply:
Yep we use the following pattern:

```rs
if let Some(chunk) = self.chunk_builder.take() {
    yield chunk;
}
```